File: Android-Native/app/src/main/java/com/kkunan/randomdogandroid/features/randomimage/data/repositories/RandomDogRepositoryImpl.kt
package com.kkunan.randomdogandroid.features.randomimage.data.repositories

import com.kkunan.randomdogandroid.common.domain.responses.LocalResponse
import com.kkunan.randomdogandroid.features.randomimage.data.datasources.RandomDogNetworkDatasource
import com.kkunan.randomdogandroid.features.randomimage.domain.entities.DogImage
import com.kkunan.randomdogandroid.features.randomimage.domain.repositories.RandomDogRepository
import javax.inject.Inject

class RandomDogRepositoryImpl @Inject constructor(private val networkDatasource: RandomDogNetworkDatasource) : RandomDogRepository {
    override suspend fun randomImage(number: Int): LocalResponse<List<DogImage>> {
        val response = networkDatasource.getRandomImage(number)
        if (response.status == "success"){
            return LocalResponse.SUCCESS(response.imageUrl?.map {
                DogImage(it)
            }?.toList() ?: emptyList())
        }
        else return LocalResponse.FAILURE("Cannot get dog images :'(")

    }
}

File: random_dog_flutter/lib/features/randomdog/data/repositories/random_dog_repository_impl.dart
import 'package:random_dog_flutter/common/data/responses/local_response.dart';
import 'package:random_dog_flutter/features/randomdog/data/datasources/network/random_image_network_datasource.dart';
import 'package:random_dog_flutter/features/randomdog/domain/entities/dog_image.dart';
import 'package:random_dog_flutter/features/randomdog/domain/repositories/random_dog_repository.dart';

class RandomDogRepositoryImpl implements RandomDogRepository {
  final RandomImageNetworkDatasource datasource;
  RandomDogRepositoryImpl(this.datasource);

  @override
  Future<LocalResponse<DogImage>> randomImage(int number) async{
    final response = await datasource.randomImage(number);
    final message = response.message;
    if(message != null){
      return LocalResponse(value: DogImage(imageUrls: message));
    } else {
      return const LocalResponse(error: "Cannot get image!");
    }
  }
}